lib/ansible/modules/identity/cyberark/cyberark_authentication.py

```diff
@@ -0,0 +1,300 @@
+
+
+def processAuthentication(module):
+
+    # Getting parameters from module
+
+    api_base_url = module.params["api_base_url"]
+    validate_certs = module.params["validate_certs"]
+    username = module.params["username"]
+    password = module.params["password"]
+    new_password = module.params["new_password"]
+    use_shared_logon_authentication = module.params[
+        "use_shared_logon_authentication"]
+    use_radius_authentication = module.params["use_radius_authentication"]
+    state = module.params["state"]
+    cyberark_session = module.params["cyberark_session"]
+
+    # if in check mode it will not perform password changes
+    if module.check_mode and new_password is not None:
+        new_password = None
+
+    # Defining initial values for open_url call
+    headers = {'Content-Type': 'application/json'}
+    payload = ""
+
+    if state == "present":  # Logon Action
+
+        # Different end_points based on the use of shared logon authentication
+        if use_shared_logon_authentication:
+
+            end_point = "/PasswordVault/WebServices/auth/Shared/RestfulAuthenticationService.svc/Logon"
+
+        else:
+
+            end_point = "/PasswordVault/WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logon"
+
+            # The payload will contain username, password
+            # and optionally use_radius_authentication and new_password
+            payload_dict = {"username": username, "password": password}
+
+            if use_radius_authentication:
+                payload_dict["useRadiusAuthentication"] = use_radius_authentication
+
+            if new_password is not None:
+                payload_dict["newPassword"] = new_password
+
+            payload = json.dumps(payload_dict)
+
+    else:  # Logoff Action
+
+        # Get values from cyberark_session already established
+        api_base_url = cyberark_session["api_base_url"]
+        validate_certs = cyberark_session["validate_certs"]
+        use_shared_logon_authentication = cyberark_session[
+            "use_shared_logon_authentication"]
+        headers["Authorization"] = cyberark_session["token"]
+
+        # Different end_points based on the use of shared logon authentication
+        if use_shared_logon_authentication:
+            end_point = "/PasswordVault/WebServices/auth/Shared/RestfulAuthenticationService.svc/Logoff"
+        else:
+            end_point = "/PasswordVault/WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logoff"
+
+    result = None
+    changed = False
+    response = None
+
+    try:
+
+        response = open_url(
+            api_base_url + end_point,
+            method="POST",
+            headers=headers,
+            data=payload,
+            validate_certs=validate_certs)
+
+    except httplib.HTTPException as e:
+
+        module.fail_json(
+            msg="end_point=%s%s" % (api_base_url, end_point),
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
```

Comment:
A couple of things, we try to avoid 'catchalls' as they don't give good error messages. We try to capture the errors after a specific action "friendly' part that suggests what was being attempted and possibly what user can do to change error to success.

Also, the 'exception' field is badly named for python, it is mostly for tracebacks and we normally append the exception string to the msg field after the 'friendly msg'